src/Dapr.PluggableComponents.AspNetCore/DaprPluggableComponentsApplication.cs

```diff
@@ -0,0 +1,179 @@
+    private readonly ConcurrentDictionary<Type, bool> registeredComponents = new ConcurrentDictionary<Type, bool>();
+    private readonly ConcurrentDictionary<Type, bool> registeredProviders = new ConcurrentDictionary<Type, bool>();
+    private readonly ConcurrentBag<DaprServiceRegistration> serviceBuilderActions = new ConcurrentBag<DaprServiceRegistration>();
+    private readonly WebApplicationBuilder webApplicationBuilder;
+
+    private DaprPluggableComponentsApplication(WebApplicationBuilder webApplicationBuilder)
+    {
+        this.webApplicationBuilder = webApplicationBuilder ?? throw new ArgumentNullException(nameof(webApplicationBuilder));
+    }
+
+    private sealed record DaprServiceRegistration(DaprPluggableComponentsServiceOptions Options, Action<DaprPluggableComponentsServiceBuilder> Callback);
+
+    public ConfigurationManager Configuration => this.webApplicationBuilder.Configuration;
+
+    public ILoggingBuilder Logging => this.webApplicationBuilder.Logging;
+
+    public IServiceCollection Services => this.webApplicationBuilder.Services;
+
+    public DaprPluggableComponentsApplication RegisterService(string socketName, Action<DaprPluggableComponentsServiceBuilder> callback)
+    {
+        return this.RegisterService(new DaprPluggableComponentsServiceOptions(socketName), callback);
+    }
+
+    public DaprPluggableComponentsApplication RegisterService(DaprPluggableComponentsServiceOptions options, Action<DaprPluggableComponentsServiceBuilder> callback)
+    {
+        this.serviceBuilderActions.Add(new DaprServiceRegistration(options, callback));
+
+        return this;
+    }
+
+    public void Run()
+    {
+        this.CreateApplication().Run();
+    }
+
+    public Task RunAsync()
+    {
+        return this.CreateApplication().RunAsync();
+    }
+
+    #region IDaprPluggableComponentsRegistrar Members
+
+    void IDaprPluggableComponentsRegistrar.RegisterAdaptor<TAdaptor>()
+        where TAdaptor : class
+    {
+        if (this.registeredAdaptors.TryAdd(typeof(TAdaptor), true))
+        {
+            this.ConfigureApplication(
+                app =>
+                {
+                    app.MapDaprPluggableComponentAdaptor<TAdaptor>();
+                });
+        }
+    }
+
+    void IDaprPluggableComponentsRegistrar.RegisterComponent<TComponent>(string socketPath) where TComponent : class
+    {
+        if (this.registeredComponents.TryAdd(typeof(TComponent), true))
+        {
+            this.ConfigureApplicationBuilder(
+                builder =>
+                {
+                    builder.Services.AddSingleton<TComponent, TComponent>();
+                });
+        }
+
+        ((IDaprPluggableComponentsRegistrar)this).RegisterComponent(socketPath, context => context.ServiceProvider.GetRequiredService<TComponent>());
+    }
+
+    private readonly DaprPluggableComponentsRegistry registry = new DaprPluggableComponentsRegistry();
```

Comment:
shouldn't this be together with the other attribute definitions above (lines 18-24).

Reply:
Fair enough. Updated.